ta/ta.mk

```diff
@@ -22,6 +22,10 @@ ifeq ($(CFG_TEE_TA_MALLOC_DEBUG),y)
 $(sm)-platform-cppflags += -DENABLE_MDBG=1
 endif
 
+ifeq ($(CFG_TA_BTI),y)
+$(sm)-platform-cppflags += -DENABLE_BTI=1
```

Comment:
Why do we need this? Same in core.mk.

Reply:
There is a possibility that user may deploy BTI enabled TA's on a non BTI enabled OP-TEE core i.e CFG_CORE_BTI=n and CFG_TA_BTI=y.  In such cases, the compiler won't generate bti instructions for C files as -mbranchprotection=bti won't be enabled. However, for assembly files in optee-core, bti instructions would be generated as BTI() gets enabled if either CFG_TA_BTI=y or CFG_CORE_BTI=y . To avoid this, ENABLE_BTI was introduced.

Reply:
Isn't it enough to let `CFG_CORE_BTI` control `mbranchprotection=bti` for OP-TEE core and vice versa for `CFG_TA_BTI`?

Reply:
mbranchprotection controls bti code in C files. For assembly files it is done via BTI MACRO.

asm.S file gets included for both optee-core and ta library. With below check, even if CFG_CORE_BTI=n, since CFG_TA_BTI is y, bti instructed gets added in the assembly files built in optee-core.
#if defined(CFG_CORE_BTI) || defined(CFG_TA_BTI) && defined(__aarch64__)
 #define BTI(...) __VA_ARGS__
 #else
 #define BTI(...)
#endif 

I may be missing something here, but CFG_TA_BTI is also included in config.mk when building op-tee core resulting in bti instructions in assembly files which is not required. It would not cause exception as pages won't be marked Guarded but adds unnecessary BTI instructions in tee.elf which is not required.

Reply:
Here's one example of a similar situation which is solved by testing for the `__KERNEL__` define.

https://github.com/OP-TEE/optee_os/blob/28bbb200bd30a638e575e01ea53207c6732d1d75/lib/libutils/ext/arch/arm/mcount_a64.S#L45

Reply:
Thanks, I will use the __KERNEL__ option